File: tests/regression/400_RUNTIME_FEATURES/430_RUNTIME/430_030_honest_benchmark/bench.py
#!/usr/bin/env python3
"""
Honest Multi-Step Benchmark - Python version
Parse and execute 3 separate strings - same as Koru
"""
import time
import sys

ITERATIONS = 10_000

def add(a, b):
    a_val = int(a)
    b_val = int(b)
    return {"sum": {"result": a_val + b_val}}

def multiply(value, by):
    val = int(value)
    by_val = int(by)
    return {"product": {"result": val * by_val}}

def subtract(value, by):
    val = int(value)
    by_val = int(by)
    return {"difference": {"result": val - by_val}}

# Three separate source strings - same as Koru
ADD_SOURCE = 'add("10", "20")'
MUL_SOURCE = 'multiply("30", "3")'
SUB_SOURCE = 'subtract("90", "5")'

def main():
    start = time.perf_counter_ns()

    final_result = 0
    for _ in range(ITERATIONS):
        # Parse and execute THREE separate strings - same as Koru
        r1 = eval(ADD_SOURCE)
        r2 = eval(MUL_SOURCE)
        r3 = eval(SUB_SOURCE)
        final_result = r3["difference"]["result"]

    elapsed_ns = time.perf_counter_ns() - start
    elapsed_ms = elapsed_ns / 1_000_000
    per_iter = elapsed_ns // ITERATIONS

    print(f"Python interpreter 3-step: {ITERATIONS} iterations", file=sys.stderr)
    print(f"Total time: {elapsed_ms:.2f} ms", file=sys.stderr)
    print(f"Per iteration: {per_iter} ns", file=sys.stderr)

    # Actual computed result
    print(f"Result: {final_result} (expected: 85)")

if __name__ == "__main__":
    main()

File: tests/regression/400_RUNTIME_FEATURES/430_RUNTIME/430_030_honest_benchmark/bench.lua
#!/usr/bin/env lua
-- Honest Multi-Step Benchmark - Lua version
-- Tests FULL interpreter: parse Lua source string -> execute
-- Apples-to-apples with Koru interpreter benchmark

local ITERATIONS = 10000

-- The functions that get called
function add(a, b)
    local a_val = tonumber(a)
    local b_val = tonumber(b)
    return { sum = { result = a_val + b_val } }
end

function multiply(value, by)
    local val = tonumber(value)
    local by_val = tonumber(by)
    return { product = { result = val * by_val } }
end

function subtract(value, by)
    local val = tonumber(value)
    local by_val = tonumber(by)
    return { difference = { result = val - by_val } }
end

-- Source strings to parse at runtime - same as Koru
local ADD_SOURCE = 'return add("10", "20")'
local MUL_SOURCE = 'return multiply("30", "3")'
local SUB_SOURCE = 'return subtract("90", "5")'

local function main()
    local start = os.clock()

    local result
    for _ = 1, ITERATIONS do
        -- Parse and execute add
        load(ADD_SOURCE)()

        -- Parse and execute multiply
        load(MUL_SOURCE)()

        -- Parse and execute subtract
        result = load(SUB_SOURCE)()
    end

    local elapsed_sec = os.clock() - start
    local elapsed_ms = elapsed_sec * 1000
    local per_iter_ns = (elapsed_sec * 1e9) / ITERATIONS

    -- Timing to stderr
    io.stderr:write(string.format("Lua interpreter 3-step: %d iterations\n", ITERATIONS))
    io.stderr:write(string.format("Total time: %.2f ms\n", elapsed_ms))
    io.stderr:write(string.format("Per iteration: %.0f ns\n", per_iter_ns))

    -- Result to stdout for verification
    print(string.format("Result: %d (expected: 85)", result.difference.result))
end

main()
